aten/src/ATen/Quantizer.h

```diff
@@ -0,0 +1,182 @@
+#pragma once
+
+#include <c10/core/QScheme.h>
+#include <c10/macros/Macros.h>
+#include <c10/util/Exception.h>
+#include <ATen/core/Tensor.h>
+
+#include <cmath>
+#include <memory>
+
+// TODO: move to c10 namespace after we
+// unified caffe2::Tensor and at::Tensor
+
+namespace at {
+
+using QTensor = Tensor;
+using RealTensor = Tensor;
+
+/**
+ * Quantizer is the class for storing all the information
+ * that's necessary to perform quantize and dequantize
+ * operation.
+ *
+ * We might have different types of quantization schemes and this is
+ * the base class for all quantizers.
+ *
+ * QTensorImpl will hold a pointer to Quantizer so that we can support
+ * different quantization schemes on Tensor.
+ *
+ * For example, the most common quantization scheme, Affine Quantization, requires
+ * scale and zero_point as parameters, we'll store scale and zero_point inside
+ * the instance and we can use it to quantize a float Tensor or dequantize a
+ * quantized Tensor.
+ *
+ * When you add new types of leaf Quantizer class, please also
+ * make sure to add a corresponding QScheme enum since
+ * they should have one to one mapping.
+ */
+struct C10_API Quantizer {
+  c10::QScheme qscheme_;
+  virtual ~Quantizer() {}
+
+  /**
+   * quantize a float Tensor into a quantized Tensor.
+   */
+  virtual QTensor quantize(RealTensor t) = 0;
+
+  /**
+   * dequantize a quantized Tensor into a float Tensor.
+   */
+  virtual RealTensor dequantize(QTensor t) = 0;
```

Comment:
should this take a dtype?  Or is the assumption that you always want to dequantize to float32 valid?

Reply:
hmm, yeah my assumption is we always want to dequantize to float, do you mean people might want float16 as well? In that case we can dequantize and then convert?
cc @raghuramank100